Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -61,69 +62,48 @@ final class ConnectionPool {
 
-        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
-            if let existing = self.connectionProviders[key] {
-                existing.stateLock.withLock {
-                    existing.state.pending += 1
-                }
+        let provider: HTTP1ConnectionProvider = self.lock.withLock {
+            if let existing = self.providers[key], existing.enqueue() {
                 return existing
             } else {
-                let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, parentPool: self)
-                self.connectionProviders[key] = http1Provider
-                http1Provider.stateLock.withLock {
-                    http1Provider.state.pending += 1
-                }
-                return http1Provider
+                let provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop, configuration: self.configuration, pool: self)
+                self.providers[key] = provider
+                return provider
             }
         }
 
-        return provider.getConnection(preference: preference)
+        return provider.getConnection(preference: preference, setupComplete: setupComplete)
     }
 
-    func release(_ connection: Connection) {
-        let connectionProvider = self.connectionProvidersLock.withLock {
-            self.connectionProviders[connection.key]
-        }
-        if let connectionProvider = connectionProvider {
-            connectionProvider.release(connection: connection)
+    func delete(_ provider: HTTP1ConnectionProvider) {
+        self.lock.withLockVoid {
+            self.providers[provider.key] = nil
         }
     }
 
-    func prepareForClose(on eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        let connectionProviders = self.connectionProvidersLock.withLock {
-            self.connectionProviders.values
+    func close(on eventLoop: EventLoop) -> EventLoopFuture<Bool> {
+        let providers = self.lock.withLock {
+            self.providers.values
         }
 
-        return EventLoopFuture<Void>.andAllComplete(connectionProviders.map { $0.prepareForClose() }, on: eventLoop)
-    }
-
-    func close(on eventLoop: EventLoop) -> EventLoopFuture<Void> {
-        let connectionProviders = self.connectionProvidersLock.withLock {
-            self.connectionProviders.values
-        }
-
-        return EventLoopFuture.andAllComplete(connectionProviders.map { $0.close() }, on: eventLoop).map {
-            self.connectionProvidersLock.withLock {
-                assert(self.connectionProviders.count == 0, "left-overs: \(self.connectionProviders)")
-            }
-        }
+        return EventLoopFuture.reduce(true, providers.map { $0.close() }, on: eventLoop) { $0 && $1 }
     }
 
     var connectionProviderCount: Int {
-        return self.connectionProvidersLock.withLock {
-            self.connectionProviders.count
+        return self.lock.withLock {
```

Comment:
do we need this `connectionProviderCount` property? Seems racy so I'd just remove it.

Reply:
we use it one place - client deinit to assert that all providers are deleted (so no decisions based on it are made)

Reply:
@artemredkin in `deinit` we don't need the lock. So we can just do

```
assert(self.providers.count == 0)
```

Reply:
👍

Reply:
done